### PR TITLE
Cred offer qr code

### PR DESCRIPTION
**Pull request by beknazaresenbek**

### CODE REVIEW

File: src/main/java/org/fiware/keycloak/VCIssuerRealmResourceProvider.java [hunk 1]
```diff
@@ -378,19 +368,70 @@ public Response getCredentialOffer(@PathParam("issuer-did") String issuerDidPara
...
+		AuthenticationManager.AuthResult authResult = getAuthResult();
+		UserSessionModel userSessionModel = getUserSessionModel();
+
+		AuthenticatedClientSessionModel clientSession = userSessionModel.
+				getAuthenticatedClientSessionByClient(
+						authResult.getClient().getId());
+		try {
+			clientSession.setNote(nonce, objectMapper.writeValueAsString(offeredCredential));
+		} catch (JsonProcessingException e) {
+			LOGGER.infof("Could not convert POJO to JSON: %s", e.getMessage());
```
Comment (pulledtim) [inline src/main/java/org/fiware/keycloak/VCIssuerRealmResourceProvider.java:382]: Is it even possible that the user sent something that we can not turn into a json? Feels more like that we did a mistake in that case.
Also I think we should use error log for this

File: src/main/java/org/fiware/keycloak/VCIssuerRealmResourceProvider.java [hunk 2]
```diff
@@ -378,19 +368,70 @@ public Response getCredentialOffer(@PathParam("issuer-did") String issuerDidPara
...
+
+	/**
+	 * Provides an OIDC4VCI compliant credentials offer 2
+	 */
+	@GET
+	@Path("{issuer-did}/credential-offer/{nonce}")
+	@Produces({ MediaType.APPLICATION_JSON })
+	public Response getCredentialOffer(@PathParam("issuer-did") String issuerDidParam,
+									   @PathParam("nonce") String nonce) {
+		LOGGER.infof("Get an offer for nonce %s", nonce);
```
Comment (pulledtim) [inline src/main/java/org/fiware/keycloak/VCIssuerRealmResourceProvider.java:404]: Might be useful to include the did in this log too

File: src/main/java/org/fiware/keycloak/VCIssuerRealmResourceProvider.java (hunk 1 above)
Comment (beknazaresenbek) [reply src/main/java/org/fiware/keycloak/VCIssuerRealmResourceProvider.java:382]: This request comes from the page, where you choose credential type and format on the Verifiable Credentials page, so there are always type and format if everything is ok. If no, there is a validity check 'getClientsOfType' at the beginning. Or what do you mean?

About the error log, yes, you're right. I'll update it

File: src/main/java/org/fiware/keycloak/VCIssuerRealmResourceProvider.java (hunk 2 above)
Comment (beknazaresenbek) [reply src/main/java/org/fiware/keycloak/VCIssuerRealmResourceProvider.java:404]: ok

File: src/main/java/org/fiware/keycloak/VCIssuerRealmResourceProvider.java [hunk 3]
```diff
@@ -378,19 +368,70 @@ public Response getCredentialOffer(@PathParam("issuer-did") String issuerDidPara
...
+
+		SupportedCredential offeredCredential;
+		try {
+			offeredCredential = objectMapper.readValue(result.getClientSession().getNote(nonce),
+					SupportedCredential.class);
+			LOGGER.infof("Creating an offer for %s - %s", offeredCredential.getType(),
+					offeredCredential.getFormat());
+			result.getClientSession().removeNote(nonce);
+		} catch (JsonProcessingException e) {
+			LOGGER.infof("Could not convert JSON to POJO: %s", e.getMessage());
```
Comment (pulledtim) [inline src/main/java/org/fiware/keycloak/VCIssuerRealmResourceProvider.java:417]: Why did not log an error here?

File: src/main/java/org/fiware/keycloak/VCIssuerRealmResourceProvider.java [hunk 4]
```diff
@@ -518,8 +564,8 @@ public Response requestCredential(@PathParam("issuer-did") String issuerDidParam
 		List<String> types = new ArrayList<>(Objects.requireNonNull(Optional.ofNullable(credentialRequestVO.getTypes())
 				.orElseGet(() -> {
 					try {
-						return objectMapper.readValue(credentialRequestVO.getType(), new TypeReference<List<String>>() {
-						});
+						return objectMapper.readValue(credentialRequestVO.getType(), new TypeReference<>() {
+                        });
```
Comment (pulledtim) [inline src/main/java/org/fiware/keycloak/VCIssuerRealmResourceProvider.java:568]: Why is the type not needed anymore?

Comment (beknazaresenbek) [reply src/main/java/org/fiware/keycloak/VCIssuerRealmResourceProvider.java:568]: Intellij said, it's redundant

**Review by pulledtim — APPROVED** [review]

**Pull request merged**